src/NUnitFramework/framework/Constraints/CollectionEquivalentConstraintResult.cs

```diff
@@ -0,0 +1,113 @@
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+
+namespace NUnit.Framework.Constraints
+{
+    /// <summary>
+    /// Provides a <see cref="ConstraintResult"/> for the <see cref="CollectionEquivalentConstraint"/>.
+    /// </summary>
+    /// <remarks>
+    /// Assumes that a <see cref="CollectionTally"/> has been created and used to compare the expected
+    /// equivalent collections.
+    /// </remarks>
+    public class CollectionEquivalentConstraintResult : ConstraintResult
+    {
+        /// <summary>
+        /// Preprocessed <see cref="CollectionTally"/> of the collections to compare for equivalence.
+        /// </summary>
+        private CollectionTally _tally;
+
+        /// <summary>
+        /// Maximum amount of elements to write to the <see cref="MessageWriter"/> if there are
+        /// extra/missing elements from the collection.
+        /// </summary>
+        private const int MaxDifferingElemsToWrite = 10;
+        
+        /// <summary>
+        /// Construct a <see cref="CollectionEquivalentConstraintResult"/> with a <see cref="CollectionTally"/>
+        /// already applied to the actual collection.
+        /// </summary>
+        /// <param name="Constraint">
+        /// Source <see cref="CollectionEquivalentConstraint"/>.
+        /// </param>
+        /// <param name="Tally">
+        /// Preprocessed <see cref="CollectionTally"/> for the collections to compare.
+        /// </param>
+        /// <param name="actual">
+        /// Collection to compare to the expected one.
+        /// </param>
+        /// <param name="isSuccess">
+        /// Whether or not the <see cref="CollectionEquivalentConstraint"/> succeeded.
+        /// </param>
+        /// <remarks>
+        /// The already processed <see cref="CollectionTally"/> is provided to reduce
+        /// unnecessary iterations over the collections to determine equivalence.
+        /// </remarks>
+        public CollectionEquivalentConstraintResult(
+            CollectionEquivalentConstraint Constraint,
+            CollectionTally Tally,
+            object actual,
+            bool isSuccess)
+            : base(Constraint, actual, isSuccess)
+        {
+            if (Tally == null)
+            {
+                throw new System.ArgumentNullException("Tally was null.");
+            }
```

Comment:
We have a `Guard.ArgumentNotNull()` method we use for cases like this - just a little tidier, and gives consistent exception messages. 😄